compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs

```diff
@@ -498,21 +500,98 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                 }
             }
+
+            self.suggest_ref_for_dbg_args(expr, span, move_span, err);
+
             if let Some(pat) = finder.pat {
-                *in_pattern = true;
-                let mut sugg = vec![(pat.span.shrink_to_lo(), "ref ".to_string())];
-                if let Some(pat) = finder.parent_pat {
-                    sugg.insert(0, (pat.span.shrink_to_lo(), "ref ".to_string()));
-                }
-                err.multipart_suggestion_verbose(
-                    "borrow this binding in the pattern to avoid moving the value",
-                    sugg,
-                    Applicability::MachineApplicable,
+                // FIXME: any better way to check this?
+                let from_std = self.infcx.tcx.sess.opts.real_rust_source_base_dir.clone().map_or(
+                    false,
+                    |root| {
+                        let file_path =
+                            match self.infcx.tcx.sess.source_map().span_to_filename(move_span) {
+                                FileName::Real(name) => {
+                                    name.clone().into_local_path().unwrap_or_default()
+                                }
+                                other => PathBuf::from(other.prefer_local().to_string()),
+                            };
+                        file_path.starts_with(&root.join("library/std/"))
+                    },
                 );
+                // it's useless to suggest inserting `ref` when the span comes from std library
+                // anyway, user can not modify std library in most cases, so let's keep it quite?
+                if !from_std {
+                    *in_pattern = true;
+                    let mut sugg = vec![(pat.span.shrink_to_lo(), "ref ".to_string())];
+                    if let Some(pat) = finder.parent_pat {
+                        sugg.insert(0, (pat.span.shrink_to_lo(), "ref ".to_string()));
+                    }
+                    err.multipart_suggestion_verbose(
+                        "borrow this binding in the pattern to avoid moving the value",
+                        sugg,
+                        Applicability::MachineApplicable,
+                    );
+                }
             }
         }
     }
 
+    // for dbg!(x) which may take onwership, suggest dbg!(&x) instead
+    // but here we actually does not checking the macro name is `dbg!`
+    // so that we may extend the scope a bit larger to cover more cases
+    fn suggest_ref_for_dbg_args(
+        &self,
+        body: &hir::Expr<'_>,
+        span: Option<Span>,
+        move_span: Span,
+        err: &mut Diag<'tcx>,
+    ) {
+        // only suggest for macro
+        if move_span.source_callsite() == move_span {
+            return;
+        }
+        let sm = self.infcx.tcx.sess.source_map();
+        let arg_code = if let Some(span) = span
+            && let Ok(code) = sm.span_to_snippet(span)
```

Comment:
I would prefer not to be using snippets for making decisions. Do we not have the expression of the value somewhere?

Reply:
Seems no, I just have a try to use ExpressionFinder to get the hir, and it only can not handle this scenario:

https://github.com/rust-lang/rust/pull/120990/files#diff-bd7f956e58a376780df31f75794e7af9b5d4043582aed6f129ae386a2b83ebdeR13-R17

the argument comes from the function parameters, since an ExpressionFinder will search in the body.